File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def cal_expire(start, month):
    y,m,d = map(int, start.split("."))
    m += month
    d -= 1
    if d < 1:
        m -= 1
        d = 28
    if m > 12:
        y += (m-1) // 12
        m = 12 if m % 12 == 0 else m % 12
    elif m < 1:
        y -= 1
        m = 12
    return f"{y}.{m:02}.{d:02}"

def is_expire(today, ex):
    return tuple(map(int, today.split("."))) > tuple(map(int, ex.split(".")))

def solution(today, terms, privacies):
    answer = []
    term = {k: int(v) for k, v in (t.split() for t in terms)}
    for idx, pri in enumerate(privacies):
        d, k = pri.split()
        if is_expire(today, cal_expire(d, term[k])):
            answer.append(idx+1)
    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.swift
import Foundation

func solution(_ today:String, _ terms:[String], _ privacies:[String]) -> [Int] {
    
    func toDays(_ date: String) -> Int {
        let d = date.split(separator: ".").map {Int($0)!}
        return d[0] * 28 * 12 + d[1] * 28 + d[2]
    }
    
    let todayDays = toDays(today)
    
    var termDict: [String: Int] = [:]
    for term in terms {
        let tmp = Array(term.split(separator: " ").map{String($0)})
        termDict[tmp[0]] = Int(tmp[1])!
    }
    
    var answer = [Int]()
    
    for i in 0..<privacies.count {
        let p = privacies[i].split(separator: " ")
        let start = String(p[0])
        let type = String(p[1])
        
        let expire = toDays(start) + termDict[type]! * 28
        
        if expire <= todayDays {
            answer.append(i+1)
        }
    }
    
    return answer
}